WIP: New Sequence data structure

mortonjt opened the pull request:
Building off of the discussions here

https://github.com/biocore/scikit-bio/pull/1157
https://github.com/biocore/scikit-bio/issues/1159
https://github.com/biocore/scikit-bio-rfcs/pull/9

Here's the tentative TODO list

- [x] port over interval tree from bxpython
- [x] Implement `Feature` class
- [x] Implement `Interval` class
- [x] implement hashtables to map between intervals and features 
- [ ] Build IntervalMetadataMixin class to inherit all of the proper methods (i.e. eq, ne, copy, and deepcopy. )
- [x] Overload `[ ]` indexing method for Sequence object to accept `Features` as input and return a new Sequence object.
- [ ] Overload `[ ]` set operation for Sequence object to accept `Features` as input and set attributes
- [x] Implement `query` method to accept an `Interval` or a `**kwargs`
- [ ] modify existing reverse complement command in `Sequence`
- [ ] Update the genbank parser

RNAer commented on skbio/sequence/_metadata.py:
```diff
@@ -0,0 +1,97 @@
+        self.interval_ids.add(ID)
+        return ID
+
+    def concat(self, x, kw='intervals'):
```
This one needs further thoughts

mortonjt commented:
Just added the `concat` and `query` methods.  So feel free to start reviewing that.

RNAer commented on skbio/sequence/_interval.py:
```diff
@@ -0,0 +1,127 @@
+        self.feature_table = defaultdict(list)
+
+    def concat(self, x, kw='location'):
+        """ Adds a feature to the metadata object.
```
I suggest renaming the function, as it is confusing with `Sequence.concat`.

mortonjt replied:
How about add?

On Fri, Feb 5, 2016 at 12:38 PM, Zech Xu notifications@github.com wrote:

> In skbio/sequence/_interval.py
> https://github.com/RNAer/scikit-bio/pull/1#discussion_r52069631:
> 
> > +
> > -    Attributes
> > -    ----------
> > -    interval_tree
> > -    feature_table
> > -    interval_table
> > -    interval_ids
> >   +
> > -    '''
> > -    def **init**(self):
> > -        self.interval_tree = IntervalTree()
> > -        # maps features to intervals and vice versa
> > -        self.feature_table = defaultdict(list)
> >   +
> > -    def concat(self, x, kw='location'):
> > -        """ Adds a feature to the metadata object.
> 
> I suggest renaming the function, as it is confusing with Sequence.concat.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/RNAer/scikit-bio/pull/1/files#r52069631.

RNAer replied:
:+1:

On Fri, Feb 5, 2016 at 1:17 PM, mortonjt notifications@github.com wrote:

> In skbio/sequence/_interval.py
> https://github.com/RNAer/scikit-bio/pull/1#discussion_r52073860:
> 
> > +
> > -    Attributes
> > -    ----------
> > -    interval_tree
> > -    feature_table
> > -    interval_table
> > -    interval_ids
> >   +
> > -    '''
> > -    def **init**(self):
> > -        self.interval_tree = IntervalTree()
> > -        # maps features to intervals and vice versa
> > -        self.feature_table = defaultdict(list)
> >   +
> > -    def concat(self, x, kw='location'):
> > -        """ Adds a feature to the metadata object.
> 
> How about add?
> … <#-558737669_>
> On Fri, Feb 5, 2016 at 12:38 PM, Zech Xu notifications@github.com
> wrote: In skbio/sequence/_interval.py <#1 (comment)
> https://github.com/RNAer/scikit-bio/pull/1#discussion_r52069631>: > + >
> - Attributes > + ---------- > + interval_tree > + feature_table > +
>   interval_table > + interval_ids > + > + ''' > + def **init**(self): > +
>   self.interval_tree = IntervalTree() > + # maps features to intervals and
>   vice versa > + self.feature_table = defaultdict(list) > + > + def
>   concat(self, x, kw='location'): > + """ Adds a feature to the metadata
>   object. I suggest renaming the function, as it is confusing with
>   Sequence.concat. — Reply to this email directly or view it on GitHub <
>   https://github.com/RNAer/scikit-bio/pull/1/files#r52069631>.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/RNAer/scikit-bio/pull/1/files#r52073860.

RNAer commented on skbio/sequence/_interval.py:
```diff
@@ -0,0 +1,127 @@
+
+
+    def query(self, **kwargs):
+        """
```
I suggest the func API is `query(self, *args, **kwargs)`, so it will provide a powerful query by combinations of locations and attributes. `args` accepts a list of location intervals and `kwargs` accepts keyword arg of attributes.

mortonjt replied:
:+1:  That's a nice way to integrate accepting intervals.

On Fri, Feb 5, 2016 at 12:41 PM, Zech Xu notifications@github.com wrote:

> In skbio/sequence/_interval.py
> https://github.com/RNAer/scikit-bio/pull/1#discussion_r52069884:
> 
> > -        # since this will be using a little more memory, since the whole
> > -        # feature object will be stored.
> > -        for interval in intervals:
> > -            self.interval_tree.add(interval[0], interval[1], x)
> > -        # TODO: The below will require careful consideration
> > -        # since this will be using a little more memory, since the whole
> > -        # feature object will be stored.
> > -        for k, v in x.items():
> > -            if x not in self.feature_table:
> > -                self.feature_table[(k, v)] = [x]
> > -            else:
> > -                self.feature_table[(k, v)].append(x)
> >   +
> >   +
> > -    def query(self, **kwargs):
> > -        """
> 
> I suggest the func API is query(self, _args, *_kwargs), so it will
> provide a powerful query by combinations of locations and attributes. args
> accepts a list of location intervals and kwargs accepts keyword arg of
> attributes.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/RNAer/scikit-bio/pull/1/files#r52069884.

mortonjt commented:
Comments addressed. 

Thinking about extending the [`__getitem__`](https://github.com/biocore/scikit-bio/blob/master/skbio/sequence/_sequence.py#L758) and adding in a `__setitem__` method to knock out points 5 and 6.  Will need some discussion before implementing.

RNAer commented on skbio/sequence/_interval.py:
```diff
@@ -0,0 +1,121 @@
+            Keyword arguments of feature name and feature value, which can
+            be passed to ``dict``.  This is used to specify the search parameters.
+            If the `location` keyword is passed, then an interval lookup will be
+            performed.
```
a pin for updating doc. what do you think of this func wrapping on 2 private funcs doing real heavy lifting? I imagine you will need query for location and query for attributes for `[ ]` overloading.

mortonjt replied:
I think making a _query_interval method is fine.  However, making a private
_query_feature is a bit redundant.  That method is already super simple.

On Fri, Feb 5, 2016 at 2:11 PM, Zech Xu notifications@github.com wrote:

> In skbio/sequence/_interval.py
> https://github.com/RNAer/scikit-bio/pull/1#discussion_r52079769:
> 
> > -        for k, v in x.items():
> > -            if x not in self.feature_table:
> > -                self.feature_table[(k, v)] = [x]
> > -            else:
> > -                self.feature_table[(k, v)].append(x)
> >   +
> >   +
> > -    def query(self, _args, *_kwargs):
> > -        """
> > -        Parameters
> > -        ----------
> > -        kwargs : dict
> > -            Keyword arguments of feature name and feature value, which can
> > -            be passed to `dict`.  This is used to specify the search parameters.
> > -            If the `location` keyword is passed, then an interval lookup will be
> > -            performed.
> 
> a pin for updating doc. what do you think of this func wrapping on 2
> private funcs doing real heavy lifting? I imagine you will need query for
> location and query for attributes for [ ] overloading.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/RNAer/scikit-bio/pull/1/files#r52079769.

RNAer replied:
right. :+1:

On Fri, Feb 5, 2016 at 2:33 PM, mortonjt notifications@github.com wrote:

> In skbio/sequence/_interval.py
> https://github.com/RNAer/scikit-bio/pull/1#discussion_r52082056:
> 
> > -        for k, v in x.items():
> > -            if x not in self.feature_table:
> > -                self.feature_table[(k, v)] = [x]
> > -            else:
> > -                self.feature_table[(k, v)].append(x)
> >   +
> >   +
> > -    def query(self, _args, *_kwargs):
> > -        """
> > -        Parameters
> > -        ----------
> > -        kwargs : dict
> > -            Keyword arguments of feature name and feature value, which can
> > -            be passed to `dict`.  This is used to specify the search parameters.
> > -            If the `location` keyword is passed, then an interval lookup will be
> > -            performed.
> 
> I think making a _query_interval method is fine. However, making a private
> _query_feature is a bit redundant. That method is already super simple.
> … <#1708571367_>
> On Fri, Feb 5, 2016 at 2:11 PM, Zech Xu notifications@github.com wrote:
> In skbio/sequence/_interval.py <#1 (comment)
> https://github.com/RNAer/scikit-bio/pull/1#discussion_r52079769>: > +
> for k, v in x.items(): > + if x not in self.feature_table: > +
> self.feature_table[(k, v)] = [x] > + else: > + self.feature_table[(k,
> v)].append(x) > + > + > + def query(self, _args, *_kwargs): > + """ > +
> Parameters > + ---------- > + kwargs : dict > + Keyword arguments of
> feature name and feature value, which can > + be passed to `dict`. This
> is used to specify the search parameters. > + If the `location` keyword is
> passed, then an interval lookup will be > + performed. a pin for updating
> doc. what do you think of this func wrapping on 2 private funcs doing real
> heavy lifting? I imagine you will need query for location and query for
> attributes for [ ] overloading. — Reply to this email directly or view it
> on GitHub https://github.com/RNAer/scikit-bio/pull/1/files#r52079769.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/RNAer/scikit-bio/pull/1/files#r52082056.

mortonjt commented:
Continuing the discussion I had with @ebolyen and @jairideout offline.

We'll try overloading the `__array__` operator, that why the `__getitem__` method won't have to be changed at all and the feature object can be automatically transformed into a boolean vector.

Now, when we were thinking about having a `__setitem__` method, we were thinking about not modifying the underlying string, rather the underlying metadata.
For example, if there is a hypothetical protein that suddenly got annotated.

``` python
protein = Feature(function='Hypothetical Protein') 
genome[protein].function = 'ABC transporter'
```

I agree, the Sequence object should be immutable.  But how should the metadata be handled?

RNAer commented:
That's fine. `Sequence` is immutable, but its metadata could. Even in the current implementation, `Sequence.metadata` is a mutable dict. the intervalmetadata should be mutable.

ebolyen commented:
:+1:

RNAer commented on skbio/sequence/_interval.py:
```diff
@@ -0,0 +1,133 @@
+    def __init__(self, interval_metadata=None, feature_table=None):
+        self._init_(self, interval_metadata=None, feature_table=None)
+
+    def _init_(self, interval_metadata=None, feature_metadata=None):
```
why not just put it in `__init__`?

mortonjt replied:
I based this on PositionalMetadataMixin

https://github.com/biocore/scikit-bio/blob/master/skbio/_base.py#L336

On Mon, Feb 8, 2016 at 1:53 PM, Zech Xu notifications@github.com wrote:

> In skbio/sequence/_interval.py
> https://github.com/RNAer/scikit-bio/pull/1#discussion_r52227737:
> 
> > -    ''' Store metadata corresponding to Features and Intervals.
> >   +
> > -    There are two underlying data structures namely
> > -    1. An IntervalTree (interval_metadata)
> > -    2. A hash table indexed by a key and value (feature_table)
> >   +
> > -    Attributes
> > -    ----------
> > -    interval_metadata
> > -    feature_table
> >   +
> > -    '''
> > -    def **init**(self, interval_metadata=None, feature_table=None):
> > -        self._init_(self, interval_metadata=None, feature_table=None)
> >   +
> > -    def _init_(self, interval_metadata=None, feature_metadata=None):
> 
> why not just put it in **init**?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/RNAer/scikit-bio/pull/1/files#r52227737.

ebolyen replied:
:+1: This gives objects implementing the mixin an obvious API to call. This ends up being more flexible than `super` in practice. Would also be good to implement `eq`, `ne`, `copy`, and `deepcopy`. The Sequence object also defers to its mixins for those operations. 

Basically anything that is wrapped in single underscores is part of the non-public (but not private) interface and exists only for dispatch. Everything else follows the standard rules.

RNAer commented on skbio/sequence/tests/test_interval.py:
```diff
@@ -0,0 +1,55 @@
+
+import unittest
+
+from skbio.sequence import IntervalMetadata, Interval, Feature
```
`ImportError: cannot import name 'IntervalMetadata'`

mortonjt replied:
Yeah.  That's out of date.  All of those tests are being moved over to test_sequence.py.

Thanks for catching that!

RNAer commented on skbio/sequence/_interval.py:
```diff
@@ -0,0 +1,74 @@
+    Attributes
+    ----------
+    interval_metadata
+    feature_metadata
```
rename this to `intervals` and `features`?

RNAer commented on skbio/sequence/_sequence.py:
```diff
@@ -522,7 +528,7 @@ def _positional_metadata_axis_len_(self):
     @stable(as_of="0.4.0")
     def __init__(self, sequence, metadata=None, positional_metadata=None,
-                 lowercase=False):
+                 lowercase=False, features=None):
```
rename features to `interval_metadata` and place it next to `positional_metadata`?

RNAer commented on skbio/sequence/_sequence.py:
```diff
@@ -1141,6 +1164,92 @@ def __deepcopy__(self, memo):
         """
         return self._copy(True, memo)
 
+    def add(self, x, *intervals):
```
rename it to `add_feature`? so that the name of `seq.add_feature()` is more straightforward. also rename `x` to `feature`.

RNAer commented on skbio/sequence/_sequence.py:
```diff
@@ -1141,6 +1164,92 @@ def __deepcopy__(self, memo):
+                 queries.append(feature)
+        return queries
+
+    def query(self, *args, **kwargs):
```
rename it to `query_interval_metadata`? kindof unwieldy. open for better name. it is not intuitive from the current name when users run `seq.query()`

RNAer commented:
Thanks for awesome work. Some suggestion on naming as I commented inline. @ebolyen , do you have opinion on the names?

The naming schema that we try to make is:

`interval_metadata` is for the whole set of sequence features with their intervals, in analogy to `positional_metadata`. `interval_metadata` has `interval`s and `feature`s. The current implementation reflects this but the naming is not in some parts. Another thing is that, although the functions are defined in IntervalMetadataMixin, but they are called upon `Sequence` object, so we'd better make it more clear those functions deal with interval metadata with a more specific name.

ebolyen commented:
@RNAer I like `interval_metadata`. One thing I would recommend is to create an object that has the `intervals` and `features` and methods and assign an instance of that to `interval_metadata`. Ideally the mixin would only be responsible for initializing and providing hooks for various builtins (`_init_`, `_eq_`, etc...).

Just like we don't have `dict` or `DataFrame` methods on `Sequence` for `metadata` and `positional_metadata` I don't think it makes sense for `Sequence` to be responsible for methods of `interval_metadata` (outside of trivial methods like `has_interval_metadata` which could be provided by the mixin).

mortonjt commented:
I'm going to close this PR and open a new one.  git is misbehaving

mortonjt commented:
Looks like the merge issues have been fixed.  Thanks @RNAer .

I think this addresses the remaining comments (save the interval/feature metadata.  Are there any other changes that need to addressed?

RNAer commented on skbio/sequence/tests/test_interval.py:
```diff
@@ -0,0 +1,81 @@
+        self.assertEqual(feats, [Feature(gene='sagB')])
+
+if __name__=='__main__':
+    pass
```
pass?

mortonjt replied:
:facepalm:
On Feb 11, 2016 6:00 PM, "Zech Xu" notifications@github.com wrote:

> In skbio/sequence/tests/test_interval.py
> https://github.com/RNAer/scikit-bio/pull/1#discussion_r52698145:
> 
> > +
> > -        feats = interval_metadata.query(gene='sagB')
> > -        self.assertEqual(feats, [Feature(gene='sagB')])
> >   +
> > -    def test_add_empty(self):
> > -        interval_metadata = IntervalMetadata()
> > -        interval_metadata.add(Feature(gene='sagA'))
> > -        feats = interval_metadata.query(gene='sagA')
> > -        self.assertEqual(feats, [Feature(gene='sagA')])
> >   +
> > -        interval_metadata.add(Feature(gene='sagB'), ())
> > -        feats = interval_metadata.query(gene='sagB')
> > -        self.assertEqual(feats, [Feature(gene='sagB')])
> >   +
> >   +if **name**=='**main**':
> > -    pass
> 
> pass?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/RNAer/scikit-bio/pull/1/files#r52698145.

RNAer commented on skbio/sequence/_interval.py:
```diff
@@ -0,0 +1,161 @@
+        self.feature_metadata[x] = list(map(_polish_interval, intervals))
+
+
+    def _query_interval(self, interval):
```
This gives duplicate if the given interval overlapping with multiple intervals of the same future. should only give one.

> > > di._query_interval((1, 5))
> > > [gene:sagA, gene:sagB, gene:sagA]

mortonjt replied:
:+1: Nice catch.

The same thing will probably also happen in _query_features.

RNAer replied:
right. and don't forget to add this case into tests when you fix it.

On Thu, Feb 11, 2016 at 10:33 PM, Jamie Morton notifications@github.com
wrote:

> In skbio/sequence/_interval.py
> https://github.com/RNAer/scikit-bio/pull/1#discussion_r52709418:
> 
> > -        for interval in intervals:
> > -            loc = _polish_interval(interval)
> > -            if loc is not None:
> > -                start, end = loc
> > -                self.interval_metadata.add(start, end, x)
> >   +
> > -        # TODO: The below will require careful consideration
> > -        # since this will be using a little more memory, since the whole
> > -        # feature object will be stored.
> > -        if x not in self.feature_metadata.keys():
> > -            self.feature_metadata[x] = []
> >   +
> > -        self.feature_metadata[x] = list(map(_polish_interval, intervals))
> >   +
> >   +
> > -    def _query_interval(self, interval):
> 
> [image: :+1:] Nice catch.
> 
> The same thing will probably also happen in _query_features.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/RNAer/scikit-bio/pull/1/files#r52709418.

RNAer commented:
given a feature, how can I get its interval(s)? is it implemented?

mortonjt commented:
Right now, that is hard-coded in the `IntervalMetadata` object.  You would have to do something as follows

``` python
intmd = IntervalMetadata(...)
intervals = intmd.feature_metadata[Feature(...)]
```

RNAer commented:
ah, missed that. this is perfectly fine. I think it is shaping up very well.

On Thu, Feb 11, 2016 at 10:45 PM, Jamie Morton notifications@github.com
wrote:

> Right now, that is hard-coded in the IntervalMetadata object. You would
> have to do something as follows
> 
> intmd = IntervalMetadata(...)
> intervals = intmd.feature_metadata[Feature(...)]
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/RNAer/scikit-bio/pull/1#issuecomment-183201527.

RNAer commented on skbio/sequence/_feature.py:
```diff
@@ -0,0 +1,52 @@
+
+
+class Feature(Mapping):
+    '''Store the metadata of a sequence feature.
```
we should enforce the presence of some keys including strand, left (open in left end), right (open in right end), type (feature type) and give them default value if not set, as these info should be required for every feature.

mortonjt replied:
If we are going to do that,  might as well hard code location no?
On Feb 11, 2016 11:00 PM, "Zech Xu" notifications@github.com wrote:

> In skbio/sequence/_feature.py
> https://github.com/RNAer/scikit-bio/pull/1#discussion_r52710617:
> 
> > +# ----------------------------------------------------------------------------
> > +# Copyright (c) 2013--, scikit-bio development team.
> > +#
> > +# Distributed under the terms of the Modified BSD License.
> > +#
> > +# The full license is in the file COPYING.txt, distributed with this software.
> > +# ----------------------------------------------------------------------------
> > +
> > +from **future** import absolute_import, division, print_function
> > +
> > +from collections import Mapping
> > +import numpy as np
> > +
> > +
> > +class Feature(Mapping):
> > -    '''Store the metadata of a sequence feature.
> 
> we should enforce the presence of some keys including strand, left (open
> in left end), right (open in right end), type (feature type) and give them
> default value if not set, as these info should be required for every
> feature.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/RNAer/scikit-bio/pull/1/files#r52710617.

RNAer replied:
let's not include location into `Feature`.

strand: controled vocabulary, '+', '-'
left & right: boolean
type: gene, CDS, exon, ncRNA, ect. open vocabulary.

On Thu, Feb 11, 2016 at 11:04 PM, Jamie Morton notifications@github.com
wrote:

> In skbio/sequence/_feature.py
> https://github.com/RNAer/scikit-bio/pull/1#discussion_r52710739:
> 
> > +# ----------------------------------------------------------------------------
> > +# Copyright (c) 2013--, scikit-bio development team.
> > +#
> > +# Distributed under the terms of the Modified BSD License.
> > +#
> > +# The full license is in the file COPYING.txt, distributed with this software.
> > +# ----------------------------------------------------------------------------
> > +
> > +from **future** import absolute_import, division, print_function
> > +
> > +from collections import Mapping
> > +import numpy as np
> > +
> > +
> > +class Feature(Mapping):
> > -    '''Store the metadata of a sequence feature.
> 
> If we are going to do that, might as well hard code location no?
> … <#2047826181_-1990970143_>
> On Feb 11, 2016 11:00 PM, "Zech Xu" notifications@github.com wrote: In
> skbio/sequence/_feature.py <#1 (comment)
> 
> ## https://github.com/RNAer/scikit-bio/pull/1#discussion_r52710617>: > +#
> 
> > +# Copyright (c) 2013--, scikit-bio development team. > +# > +#
> > Distributed under the terms of the Modified BSD License. > +# > +# The full
> > 
> > ## license is in the file COPYING.txt, distributed with this software. > +#
> > - > +from **future** import absolute_import, division, print_function > +
> >   +from collections import Mapping > +import numpy as np > + > + > +class
> >   Feature(Mapping): > + '''Store the metadata of a sequence feature. we
> >   should enforce the presence of some keys including strand, left (open in
> >   left end), right (open in right end), type (feature type) and give them
> >   default value if not set, as these info should be required for every
> >   feature. — Reply to this email directly or view it on GitHub <
> >   https://github.com/RNAer/scikit-bio/pull/1/files#r52710617>.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/RNAer/scikit-bio/pull/1/files#r52710739.

mortonjt commented:
I'm a little hesitant to enforce a controlled vocab - what if we are looking at transcripts or proteins?  

Will require further discussion.

RNAer commented:
ok. let's not do that. I was mostly thinking of DNA/RNA.

On Fri, Feb 12, 2016 at 11:21 AM, Jamie Morton notifications@github.com
wrote:

> I'm a little hesitant to enforce a controlled vocab - what if we are
> looking at transcripts or proteins?
> 
> Will require further discussion.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/RNAer/scikit-bio/pull/1#issuecomment-183456474.

RNAer commented on skbio/sequence/_interval.py:
```diff
@@ -0,0 +1,161 @@
+                    self.intervals.add(start, end, k)
+            self.features = features
+
+    def add(self, x, *intervals):
```
rename `x` to `feature`?

RNAer commented:
notice the test failed after you corrected `pass` to tests.

RNAer commented:
gonna merge it as discussed. the failings are due to PEP8. we can't fix them later.

mortonjt commented:
Linking this PR here make it easier to find: https://github.com/RNAer/scikit-bio/pull/2
